blocks/bricks/recurrent.py

```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+        for level in range(1, depth):
+            for property_ in exposed_parameters:
+                setattr(self.apply,
+                        property_,
+                        getattr(self.apply, property_) +
+                        self.suffixes(getattr(transitions[level].apply,
+                                              property_),
+                                      level)
+                        )
+
+        # place mask at end because it has a default value (None)
+        # and therefor should come after arguments that may come us
+        # unnamed arguments
+        if "mask" in transitions[0].apply.sequences:
+            self.apply.sequences.append("mask")
+
+        # add context
+        self.apply.contexts = list(set(
+            sum([transition.apply.contexts for transition in transitions], [])
+        ))
+
+        # sum up all the arguments we exepct to see in a call to a transition
+        # apply method, anything else is a recursion control
+        self.transition_args = set(self.apply.sequences +
+                                   self.apply.states +
+                                   self.apply.contexts)
+
+    def normal_inputs(self, level):
+        return [name for name in self.transitions[level].apply.sequences
+                if name != 'mask']
+
+    def _push_allocation_config(self):
+        # Configure the forks that connect the "states" element in the `states`
+        # of one layer to the elements in the `sequences` of the next layer,
+        # excluding "mask".
+        # This involves `get_dim` requests
+        # to the transitions. To make sure that it answers
+        # correctly we should finish its configuration first.
+        for transition in self.transitions:
+            transition.push_allocation_config()
+
+        for level, fork in enumerate(self.forks):
+            fork.input_dim = self.transitions[level].get_dim(self.states_name)
+            fork.output_dims = self.transitions[level + 1].get_dims(
+                fork.output_names)
+
+    def do_apply(self, *args, **kwargs):
+        """Apply the stack of transitions.
+
+        This is the undecorated implementation of the apply method.
+        It is separated from the decorated apply method in order to allow
+        usage of different docrations (wrappers) to be used.
+
+        Parameters
+        ----------
+        See docstring of the class for arguments appearing in
+        self.apply.sequences, self.apply.states, self.apply.contexts
+        All arguments values are of type :class:`~tensor.TensorVariable`.
+
+        In addition the `iterate`, `reverse`, `return_initial_states` or
+        any other argument defined in `recurrent_apply` wrapper.
+
+        Returns
+        -------
+        The outputs of all transitions as defined in `self.apply.outputs`
+        All return values are of type :class:`~tensor.TensorVariable`.
+
+        """
+        nargs = len(args)
```

Comment:
It would be cool to support positional arguments here, but then contexts should also be supported. For the first version supporting only keyword arguments would be sufficient I would say.